File: core/src/main/java/io/github/bindglam/neko/NekoPluginImpl.java
package io.github.bindglam.neko;

import io.github.bindglam.neko.manager.*;
import lombok.Getter;
import lombok.experimental.Accessors;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.server.ServerLoadEvent;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.List;

public final class NekoPluginImpl extends JavaPlugin implements NekoPlugin {
    @Getter @Accessors(fluent = true)
    private final RegistryManagerImpl registryManager = new RegistryManagerImpl();
    @Getter @Accessors(fluent = true)
    private final ContentManagerImpl contentManager = new ContentManagerImpl();
    @Getter @Accessors(fluent = true)
    private final ResourcePackManagerImpl resourcePackManager = new ResourcePackManagerImpl();

    private final CommandManager commandManager = new CommandManager();

    private final List<Managerial> managers = List.of(
            registryManager,
            contentManager,
            resourcePackManager,
            commandManager
    );

    @Override
    public void onEnable() {
        Neko.registerPlugin(this);

        Context context = new Context(this);
        managers.forEach(manager -> manager.preload(context));

        getServer().getPluginManager().registerEvents(new ServerLoadListener(), this);
    }

    @Override
    public void onDisable() {
        Context context = new Context(this);
        managers.forEach(manager -> manager.end(context));

        Neko.unregisterPlugin();
    }

    @Override
    public void reload() {
        List<Reloadable> reloadableList = managers.stream()
                .filter(manager -> manager instanceof Reloadable)
                .map(manager -> (Reloadable) manager)
                .toList();

        Context context = new Context(this);
        reloadableList.forEach(reloadable -> reloadable.end(context));
        reloadableList.forEach(reloadable -> reloadable.preload(context));
        reloadableList.forEach(reloadable -> reloadable.start(context));
    }

    private final class ServerLoadListener implements Listener {
        @EventHandler
        public void onServerLoad(ServerLoadEvent event) {
            if (event.getType() == ServerLoadEvent.LoadType.RELOAD) {
                return;
            }

            Context context = new Context(NekoPluginImpl.this);
            managers.forEach(manager -> manager.start(context));
        }
    }
}

File: core/src/main/kotlin/io/github/bindglam/neko/NekoPluginImpl.kt
package io.github.bindglam.neko

import io.github.bindglam.neko.manager.CommandManager
import io.github.bindglam.neko.manager.ContentManager
import io.github.bindglam.neko.manager.ContentManagerImpl
import io.github.bindglam.neko.manager.Context
import io.github.bindglam.neko.manager.RegistryManagerImpl
import io.github.bindglam.neko.manager.ResourcePackManager
import io.github.bindglam.neko.manager.ResourcePackManagerImpl
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.server.ServerLoadEvent
import org.bukkit.plugin.java.JavaPlugin

class NekoPluginImpl : JavaPlugin(), NekoPlugin {
    private val managers = listOf(
        ContentManagerImpl,
        RegistryManagerImpl,
        ResourcePackManagerImpl,
        CommandManager
    )

    override fun onEnable() {
        Neko.registerPlugin(this)

        managers.forEach { it.preload(Context(this)) }

        server.pluginManager.registerEvents(object : Listener {
            @EventHandler
            fun ServerLoadEvent.onStartup() {
                if(type == ServerLoadEvent.LoadType.RELOAD) return

                managers.forEach { it.start(Context(this@NekoPluginImpl)) }
            }
        }, this)
    }

    override fun onDisable() {
        managers.forEach { it.end(Context(this)) }

        Neko.unregisterPlugin()
    }

    override fun registryManager() = RegistryManagerImpl
    override fun contentManager() = ContentManagerImpl
    override fun resourcePackManager() = ResourcePackManagerImpl
}